File: my-folder/0680-valid-palindrome-ii/solution.py
class Solution:
    
    def validPalindrome(self, s: str) -> bool:
        l = 0
        r = len(s) - 1

        def is_palindrome(substring, l, r):
            while l < r:
                if s[l] != s[r]:
                    return False
                l += 1
                r -= 1
            return True
        
        while l < r:
            if s[l] == s[r]:
                l += 1
                r -= 1
            else:
                skip_left = is_palindrome(s, l + 1, r)
                skip_right = is_palindrome(s, l, r - 1)
                return skip_left or skip_right
        return True
                

File: my-folder/0680-valid-palindrome-ii/solution.java
class Solution {
    public boolean validPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;
        while(left < right){
            // when we see a mismatched character we update flag to false and move it to next character
            if(s.charAt(left) != s.charAt(right)){
                return isPalindrome(left + 1, right, s) || isPalindrome(left, right - 1, s);
            }
            left += 1;
            right -= 1;
        }
        return true;
    }


    private boolean isPalindrome(int left, int right, String s){
        while(left  < right){
            if(s.charAt(left) != s.charAt(right)){
                return false;
            }
            left += 1;
            right -= 1;
        }
        return true;
    }
      
}
